### PR TITLE
ROS1 publishing support

### PR DESCRIPTION
**Pull request by jhurliman**
This adds publishing support to the `@foxglove/ros1` package. The primary new methods are `RosNode#advertise` and `RosNode#publish`. Advertised topics are described by topic, dataType, and any combination of messageDefinition, messageDefinitionText, and md5sum. This allows topics where we have the ROS msgdef to be republished, or new message types can by synthesized at runtime.

## TODO

 - [x] rosmaster server implementation
 - [x] integration testing using internal rosmaster server

### CODE REVIEW

**Comment by defunctzombie** [comment]
https://github.com/foxglove/studio/pull/981

File: licenses.test.ts [hunk 1]
```diff
@@ -28,27 +28,34 @@ jest.setTimeout(30 * 1000);
...
-            return;
-          }
-          resolve(modules);
-        },
-      );
-    });
-
-    // eslint-disable-next-line
-    console.log((checker as any).asSummary(packages));
+    // await printLicenses();
```
Comment (jtbandes) [inline licenses.test.ts:31]: what's happening with this?

File: licenses.test.ts [hunk 2]
```diff
@@ -28,27 +28,34 @@ jest.setTimeout(30 * 1000);
...
+    await checkLicenses(ALLOWED_LICENSES.join(";"));
   });
 });
+
+// async function printLicenses() {
```
Comment (defunctzombie) [inline licenses.test.ts:36]: Dead code?

File: licenses.test.ts (hunk 1 above)
Comment (jhurliman) [reply licenses.test.ts:31]: Nothing, it's commented out

File: packages/ros1/src/Publication.ts [hunk 3]
```diff
@@ -11,44 +15,95 @@ type SubscriberStats = [number, number, number, number, 0];
 // e.g. [2, "/listener", "o", "TCPROS", "/chatter", true, "TCPROS connection on port 55878 to [127.0.0.1:44273 on socket 7]"]
 type SubscriberInfo = [number, string, "o", string, string, number, string];
 
-export class Publication {
+export class Publication extends EventEmitter {
```
Comment (jtbandes) [inline packages/ros1/src/Publication.ts:18]: Is EventEmitter interface used? Doesn't look like it has been used in the class body – if it is used, please specify a type map here so the .on() methods are properly typed.

File: packages/ros1/src/RosMaster.ts [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+  for (let i = 0; i < args.length; i++) {
+    if (expected[i] !== "*" && typeof args[i] !== expected[i]) {
+      return new Error(`Expected "${expected[i]}" for arg ${i}, got "${typeof args[i]}"`);
+    }
+  }
+
+  return undefined;
+}
+
+export class RosMaster extends EventEmitter {
```
Comment (jtbandes) [inline packages/ros1/src/RosMaster.ts:28]: Comment - what is this used for? How is it used?

File: packages/ros1/src/RosMaster.ts [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+    this._server.close();
+  }
+
+  url(): string | undefined {
+    return this._url;
+  }
+
+  // <http://wiki.ros.org/ROS/Master_API> handlers
+
+  registerService = (_: string, args: XmlRpcValue[]): Promise<RosXmlRpcResponse> => {
```
Comment (jtbandes) [inline packages/ros1/src/RosMaster.ts:84]: what are these dangling string params? Can aid readability if the name is specified, so we know what is being ignored

File: packages/ros1-turtlesim-test/src/turtlesim.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+import { RosNode } from "@foxglove/ros1";
+import {
+  getEnvVar,
+  getHostname,
+  getNetworkInterfaces,
+  getPid,
+  TcpServerNode,
+  TcpSocketNode,
+} from "@foxglove/ros1/src/nodejs";
```
Comment (defunctzombie) [inline packages/ros1-turtlesim-test/src/turtlesim.ts:13]: I think it would be nice for the node interfaces to be separate package so users are not importing from "src" folder which is a bit of an anti-pattern? Can be follow-up PR since I know thats the status quo going into this PR.

File: packages/ros1/src/TcpPublisher.ts [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+      topic: string,
+      connectionId: number,
+      destinationCallerId: string,
+      client: Client,
+    ) => void,
+  ): this;
+  on(event: string, listener: ListenerFn): this;
+}
+
+// Implements a subscriber for the TCPROS transport. The actual TCP transport is
```
Comment (jtbandes) [inline packages/ros1/src/TcpPublisher.ts:27]: publisher?

File: packages/ros1/src/nodejs/roscore.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (defunctzombie) [inline packages/ros1/src/nodejs/roscore.ts:1]: Should this live in a bin folder? or maybe a separate ros1-core package? I like the idea of it being clear what are binaries and what are libraries.

File: packages/ros1/src/TcpPublisher.ts [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+import { EventEmitter, ListenerFn } from "eventemitter3";
+
+import { Client } from "./Client";
+import { LoggerService } from "./LoggerService";
+import { Publication } from "./Publication";
+import { Publisher } from "./Publisher";
+import { PublicationLookup, TcpClient } from "./TcpClient";
+import { TcpAddress, TcpServer, TcpSocket } from "./TcpTypes";
+
+export declare interface TcpPublisher {
```
Comment (jtbandes) [inline packages/ros1/src/TcpPublisher.ts:14]: This should be done using the EventEmitter generic param instead of a separate interface.

File: packages/ros1/src/rosmsg.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import { RosMsgDefinition } from "rosbag";
+
+export function rosMsgDefinitionText(msgDefs: RosMsgDefinition[]): string {
+  let output = "";
```
Comment (defunctzombie) [inline packages/ros1/src/rosmsg.ts:8]: random alternative suggestion: build up an array of strings and then `.join('\n')` at the end (can ignore if you want or do in another PR)

File: packages/ros1/src/mock/MockHttpServer.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+  private _listening = false;
+  private _port?: number;
+  private _hostname?: string;
+
+  url(): string | undefined {
+    if (!this._listening) {
+      return undefined;
+    }
+    return `http://${this._hostname ?? "127.0.0.1"}:${this._port ?? 11311}/`;
```
Comment (jtbandes) [inline packages/ros1/src/mock/MockHttpServer.ts:18]: seems like these defaults (at least the port) don't belong in a class generically called MockHttpServer – can we just require them to be passed in at construction?

File: packages/ros1/src/rosmsg.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+  for (let i = 0; i < msgDefs.length; i++) {
+    const msgDef = msgDefs[i] as RosMsgDefinition;
+    const constants = msgDef.definitions.filter((d) => d.isConstant);
```
Comment (defunctzombie) [inline packages/ros1/src/rosmsg.ts:11]: please avoid single letter vars - even in these cases - I'm a fan of words :D

File: packages/xmlrpc/src/XmlRpcServer.ts [hunk 13]
```diff
@@ -8,7 +8,11 @@ import { Deserializer } from "./Deserializer";
 import { HttpRequest, HttpResponse, HttpServer } from "./HttpTypes";
 import { serializeFault, serializeMethodResponse } from "./Serializer";
 import { XmlRpcFault } from "./XmlRpcFault";
-import { XmlRpcMethodHandler } from "./XmlRpcTypes";
+import { XmlRpcMethodHandler, XmlRpcValue } from "./XmlRpcTypes";
+
+const APPLICATION_ERROR = -32500;
```
Comment (defunctzombie) [inline packages/xmlrpc/src/XmlRpcServer.ts:13]: Maybe an errors enum?

File: packages/ros1/src/RosNode.test.ts [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+    const args = await paramUpdated;
+    expect(args.callerId).toEqual("/nodeA");
+    expect(args.key).toEqual("a");
+    expect(args.value).toEqual(42);
+    expect(args.prevValue).toBeUndefined();
+
+    nodeB.shutdown();
+    nodeA.shutdown();
+    await new Promise((r) => setTimeout(r, 100));
+    rosMaster.close();
```
Comment (jtbandes) [inline packages/ros1/src/RosNode.test.ts:88]: should the close function be async so we can actually wait for teardown?

File: packages/xmlrpc/src/XmlRpcServer.ts [hunk 15]
```diff
@@ -41,26 +45,63 @@ export class XmlRpcServer {
...
+    if (methodName === "system.multicall") {
+      if (!Array.isArray(args) || args.length !== 1 || !Array.isArray(args[0])) {
+        body = serializeFault(new XmlRpcFault("Invalid system.multicall", INVALID_PARAMS_ERROR));
+      } else {
+        const calls = args[0] as { methodName: string; params: XmlRpcValue[] }[];
+        const responses = await Promise.all(
+          calls.map((c) => this._methodCallHandler(c.methodName, c.params)),
         );
+        const res: XmlRpcValue[] = responses.map((r) => {
+          if (r instanceof XmlRpcFault) {
```
Comment (defunctzombie) [inline packages/xmlrpc/src/XmlRpcServer.ts:67]: Please avoid single letter var names - I have a preference for words when reading the code /shrug

File: packages/ros1/src/nodejs/roscore.ts [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+import { RosMaster } from "../RosMaster";
+import { RosNode } from "../RosNode";
+
+async function main() {
+  const hostname = RosNode.GetRosHostname(getEnvVar, getHostname, getNetworkInterfaces);
+  const port = 11311;
+  const httpServer = new HttpServerNodejs();
+  const rosMaster = new RosMaster(httpServer);
+
+  await rosMaster.start(hostname, port);
```
Comment (jtbandes) [inline packages/ros1/src/nodejs/roscore.ts:17]: should we gracefully close this in an atexit handler?

File: packages/ros1/src/TcpPublisher.ts [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+    if (!this._shutdown) {
+      this._log?.warn?.(`tcp publisher error: ${err}`);
+    }
+  };
+}
+
+async function socketRemoteAddress(socket: TcpSocket): Promise<TcpAddress | undefined> {
+  try {
+    return await socket.remoteAddress();
+  } catch {
```
Comment (defunctzombie) [inline packages/ros1/src/TcpPublisher.ts:151]: Error thrown away? this could make debugging hard when it fails - why is the error discarded here? I'm tempted to say it should not be.

File: packages/ros1/src/rosmsg.ts [hunk 18]
```diff
@@ -0,0 +1,32 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import { RosMsgDefinition } from "rosbag";
+
+export function rosMsgDefinitionText(msgDefs: RosMsgDefinition[]): string {
```
Comment (jtbandes) [inline packages/ros1/src/rosmsg.ts:7]: Needs a comment. Also needs unit tests.

File: packages/ros1/src/TcpPublisher.ts [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+export class TcpPublisher extends EventEmitter implements Publisher {
+  private _server: TcpServer;
+  private _nodeName: string;
+  private _getConnectionId: () => number;
+  private _getPublication: PublicationLookup;
+  private _pendingClients = new Map<number, TcpClient>();
+  private _shutdown = false;
+  private _log?: LoggerService;
+
+  constructor(
```
Comment (defunctzombie) [inline packages/ros1/src/TcpPublisher.ts:42]: Please use an options object - this is too many args to pass individually - makes call site reading hard(er).

File: packages/xmlrpc/src/XmlRpcServer.ts (hunk 13 above)
Comment (jtbandes) [inline packages/xmlrpc/src/XmlRpcServer.ts:13]: Is there a link we can reference for where these standard values come from?

File: packages/ros1/src/TcpPublisher.ts (hunk 7 above)
Comment (defunctzombie) [inline packages/ros1/src/TcpPublisher.ts:27]: I think we are using /** */ style comments so vscode can provide the documentation hints.

File: licenses.test.ts (hunk 1 above)
Comment (jtbandes) [reply licenses.test.ts:31]: Ok, please delete it then.

File: packages/ros1/src/rosmsg.ts (hunk 12 above)
Comment (jtbandes) [reply packages/ros1/src/rosmsg.ts:11]: the zero-letter variable name: `filter(({ isConstant }) => isConstant)`

File: packages/ros1/src/TcpConnection.ts [hunk 20]
```diff
@@ -144,7 +144,8 @@ export class TcpConnection extends EventEmitter implements Connection {
   }
 
   toString(): string {
-    return `tcpros://${this._address}:${this._port}`;
+    const host = this._address.includes(":") ? `[${this._address}]` : this._address;
```
Comment (defunctzombie) [inline packages/ros1/src/TcpConnection.ts:147]: This could use a comment - I'm not following why this syntax is important in this case - I'm guessing something to do with tcpros:// assumptions but it would be nice to document.

File: packages/ros1/src/RosNode.ts [hunk 21]
```diff
@@ -383,9 +462,13 @@ export class RosNode extends EventEmitter {
     return { port: protocol[2] as number, address: protocol[1] as string };
   }
 
-  private _newConnectionId(): number {
+  private _newConnectionId = (): number => {
```
Comment (defunctzombie) [inline packages/ros1/src/RosNode.ts:465]: why did these change to property assignments rather than class functions? Does this make a new function on every instance of the class rather than making the function once? Or does it make no difference to the number of functions created?

File: packages/ros1/src/RosNode.ts [hunk 22]
```diff
@@ -429,6 +566,34 @@ export class RosNode extends EventEmitter {
...
+    const callerApi = this._callerApi();
+
+    const [status, msg, subscribers] = await this.rosMasterClient.registerPublisher(
+      this.name,
+      publication.name,
+      publication.dataType,
+      callerApi,
+    );
+
+    if (status !== 1) {
```
Comment (defunctzombie) [inline packages/ros1/src/RosNode.ts:583]: magic number `1` could use an enum home

File: packages/ros1/src/RosNode.ts [hunk 23]
```diff
@@ -429,6 +566,34 @@ export class RosNode extends EventEmitter {
...
+      throw new Error(`registerPublisher() failed. status=${status}, msg="${msg}"`);
+    }
+
+    this._log?.debug?.(`registered publisher for ${publication.name} (${publication.dataType})`);
+    if (!Array.isArray(subscribers)) {
+      throw new Error(
```
Comment (defunctzombie) [inline packages/ros1/src/RosNode.ts:589]: RosNode is an event emitter - does it make sense for this throw to instead emit an error? Then the creator of RosNode can handle "errors" it emits? Or is this call (_registerPublisher) the direct result of a user invoked function rather than background read/write?

File: packages/ros1/src/Subscription.ts [hunk 24]
```diff
@@ -47,13 +47,15 @@ export class Subscription extends EventEmitter {
   readonly name: string;
   readonly md5sum: string;
   readonly dataType: string;
+  readonly tcpNoDelay: boolean;
   private _publishers = new Map<number, PublisherLink>();
 
-  constructor(name: string, md5sum: string, dataType: string) {
+  constructor(name: string, md5sum: string, dataType: string, tcpNoDelay: boolean) {
```
Comment (defunctzombie) [inline packages/ros1/src/Subscription.ts:53]: Too many args - should be an options object now.

File: packages/ros1/src/TcpClient.ts [hunk 25]
```diff
@@ -0,0 +1,220 @@
...
+  private _nodeName: string;
+  private _connected = true;
+  private _receivedHeader = false;
+  private _transportInfo: string;
+  private _stats: ClientStats = { bytesSent: 0, bytesReceived: 0, messagesSent: 0 };
+  private _getPublication: PublicationLookup;
+  private _log?: LoggerService;
+  private _transformer: RosTcpMessageStream;
+
+  constructor(
```
Comment (defunctzombie) [inline packages/ros1/src/TcpClient.ts:35]: Too many args - should be an options object now.

File: packages/ros1/src/TcpClient.ts [hunk 26]
```diff
@@ -0,0 +1,220 @@
...
+    socket: TcpSocket,
+    address: string,
+    port: number,
+    nodeName: string,
+    getPublication: PublicationLookup,
```
Comment (defunctzombie) [inline packages/ros1/src/TcpClient.ts:40]: This option feels weird to me - could the caller announce they are publishing things be invoking a function on the TcpClient rather than having the client call out to functions passed in? There are times when I think this works - but here it feels like there is a more natural alternative.

File: packages/ros1/src/nodejs/roscore.ts (hunk 8 above)
Comment (jhurliman) [reply packages/ros1/src/nodejs/roscore.ts:1]: This isn't a real binary. The thing living in a bin folder would look like this:

```
#!/usr/bin/env node
require('./roscore.js')
```

Comment (defunctzombie) [reply packages/ros1/src/nodejs/roscore.ts:1]: Its also not a library tho (doesn't export anything) - I dunno - just feels to me like module organization could be clearer here - or just make a ros1-core package for now that has the binary. I think its absolutely ok to separate out library from specific binaries you might run. Maybe this gets an arg parser later - now your ros1 package depends on an arg parser when no real reason for that. These are good examples for readers.

File: packages/ros1/src/RosNode.test.ts (hunk 14 above)
Comment (jhurliman) [reply packages/ros1/src/RosNode.test.ts:88]: That would be nice, but it would require rewriting a decent amount of code and should be a separate ticket+PR.

File: packages/ros1/src/nodejs/roscore.ts (hunk 16 above)
Comment (jhurliman) [reply packages/ros1/src/nodejs/roscore.ts:17]: What would the graceful closing do? Killing the process will already close any listening sockets. It could broadcast out empty publisher lists to all subscribers but I think that deviates from the `ros_comm` implementation.

File: packages/ros1/src/TcpPublisher.ts (hunk 7 above)
Comment (jhurliman) [reply packages/ros1/src/TcpPublisher.ts:27]: This bug is being fixed https://github.com/microsoft/tsdoc/issues/160

File: packages/ros1/src/TcpConnection.ts (hunk 20 above)
Comment (jhurliman) [reply packages/ros1/src/TcpConnection.ts:147]: Adding a comment. This is required by RFC2732 for printing IPv6 addresses.

File: packages/ros1/src/TcpPublisher.ts (hunk 7 above)
Comment (defunctzombie) [reply packages/ros1/src/TcpPublisher.ts:27]: My last reading of that issue did not give me the sense it is being fixed. Thats an open discussion. Have any of the ts core devs chimed in on it and indicated either way?

File: packages/ros1/src/RosNode.ts (hunk 21 above)
Comment (jhurliman) [reply packages/ros1/src/RosNode.ts:465]: It was required to make `getConnectionId: this._newConnectionId,` above work. I assume this makes a new function on every instance of the class unless there is some V8 optimization under the hood.

File: packages/ros1/src/RosNode.ts (hunk 23 above)
Comment (jhurliman) [reply packages/ros1/src/RosNode.ts:589]: This is only called directly from the user-invoked `advertise()` method so we want it to throw.

File: packages/ros1/src/RosNode.ts (hunk 21 above)
Comment (defunctzombie) [reply packages/ros1/src/RosNode.ts:465]: Oh I see - cause it is bound - I dunno - feels fragile - getConnectionId should maybe do ` = () => this._newConnectionId` so it is clear what is going on.

File: packages/ros1/src/TcpClient.ts (hunk 26 above)
Comment (jhurliman) [reply packages/ros1/src/TcpClient.ts:40]: I'm not following the suggestion. `const pub = this._getPublication(topic);` needs to be called when a certain message is parsed on the TCP socket handler, or in some other place for a different transport. It's not something the caller controls the timing of. Maybe a code example in a separate PR would help.

File: packages/ros1/src/nodejs/roscore.ts (hunk 8 above)
Comment (jhurliman) [reply packages/ros1/src/nodejs/roscore.ts:1]: The TypeScript creators disagree with you that this is not a library. See:

https://github.com/microsoft/TypeScript/blob/master/lib/tsc.js
https://github.com/microsoft/TypeScript/blob/master/bin/tsc

That repo is also an example of disagreeing with the "no mixing binaries and libraries in one package" rule. Since there is other cleanup that needs to happen with package layout here like removing `/src/` from import statements I imagine we will follow up with one or more reorg PRs where this can be addressed.

File: packages/ros1/src/TcpPublisher.ts (hunk 7 above)
Comment (jhurliman) [reply packages/ros1/src/TcpPublisher.ts:27]: My reading of that and the linked proposal PR suggested it was moving forward, but maybe that's overly optimistic. This style of comment is used extensively throughout our packages, I'll look at options for automating a conversion to `/** */`
